azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSOutputStream.java

```diff
@@ -0,0 +1,121 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.azure.adlsv2;
+
+import com.azure.storage.common.ParallelTransferOptions;
+import com.azure.storage.file.datalake.DataLakeFileClient;
+import com.azure.storage.file.datalake.options.DataLakeFileOutputStreamOptions;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.io.FileIOMetricsContext;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.metrics.Counter;
+import org.apache.iceberg.metrics.MetricsContext;
+import org.apache.iceberg.metrics.MetricsContext.Unit;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ADLSOutputStream extends PositionOutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(ADLSOutputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final DataLakeFileClient fileClient;
+  private final AzureProperties azureProperties;
+
+  private OutputStream stream;
+
+  private final Counter writeBytes;
+  private final Counter writeOperations;
+
+  private long pos;
+  private boolean closed;
+
+  ADLSOutputStream(
+      DataLakeFileClient fileClient, AzureProperties azureProperties, MetricsContext metrics)
+      throws IOException {
+    this.fileClient = fileClient;
+    this.azureProperties = azureProperties;
+
+    this.createStack = Thread.currentThread().getStackTrace();
+
+    this.writeBytes = metrics.counter(FileIOMetricsContext.WRITE_BYTES, Unit.BYTES);
+    this.writeOperations = metrics.counter(FileIOMetricsContext.WRITE_OPERATIONS);
+
+    openStream();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void flush() throws IOException {
+    stream.flush();
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    stream.write(b);
+    pos += 1;
+    writeBytes.increment();
+    writeOperations.increment();
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    stream.write(b, off, len);
+    pos += len;
+    writeBytes.increment(len);
+    writeOperations.increment();
+  }
+
+  private void openStream() {
+    DataLakeFileOutputStreamOptions options = new DataLakeFileOutputStreamOptions();
+    ParallelTransferOptions transferOptions = new ParallelTransferOptions();
+    azureProperties.adlsWriteBlockSize().ifPresent(transferOptions::setBlockSizeLong);
+    this.stream = new BufferedOutputStream(fileClient.getOutputStream(options));
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (closed) {
+      return;
+    }
+
+    super.close();
+    this.closed = true;
+    stream.close();
```

Comment:
I wonder if this should be in a try-catch in case closing fails

Reply:
also do we need a null check against `stream`?

Reply:
I updated both the input and output streams to handle these cases

Reply:
we're probably ok not wrapping this in a try-catch. I saw that one FileIO impl had a try-finally and mistakenly thought that we might want to wrap this in a try-catch